File: leet/src/Q_1_10/Q4/Q4.go
package main

import "math"

//执行用时 :20 ms, 在所有 Go 提交中击败了83.49%的用户
//内存消耗 :5.5 MB, 在所有 Go 提交中击败了92.74%的用户
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {

	A := nums1
	B := nums2
	if len(nums1) > len(nums2) {
		B = nums1
		A = nums2
	}
	m := len(A)
	n := len(B)
	iMin := 0
	iMax := m
	halfLen := (m + n + 1) / 2
	for {
		if iMin > iMax {
			break
		}

		i := (iMin + iMax) / 2
		j := halfLen - i
		if i < iMax && B[j-1] > A[i] {
			iMin = i + 1
		} else if i > iMin && A[i-1] > B[j] {
			iMax = i - 1
		} else {
			maxLeft := float64(0)
			if i == 0 {
				maxLeft = float64(B[j-1])
			} else if j == 0 {
				maxLeft = float64(A[i-1])
			} else {
				maxLeft = math.Max(float64(A[i-1]), float64(B[j-1]))
			}

			if (m+n)%2 == 1 {
				return float64(maxLeft)
			}
			minRight := float64(0)
			if i == m {
				minRight = float64(B[j])
			} else if j == n {
				minRight = float64(A[i])
			} else {
				minRight = math.Min(float64(A[i]), float64(B[j]))
			}
			return (maxLeft + minRight) / 2
		}

	}
	return 0
}

File: leet/src/Q_1_10/Q4/Q4.py
class Solution(object):
# 执行用时 :96 ms, 在所有 Python 提交中击败了88.73%的用户
# 内存消耗 :11.9 MB, 在所有 Python 提交中击败了25.79%的用户

    def findMedianSortedArrays(self, nums1, nums2):
        A = nums1
        B = nums2

        if len(nums1) > len(nums2):
            B = nums1
            A = nums2

        m = len(A)
        n = len(B)

        iMin = 0
        iMax = m

        halfLen = (m + n + 1) / 2
        while iMin <= iMax:
            i = (iMin + iMax) / 2
            j = halfLen - i
            if i < iMax and B[int(j - 1)] > A[int(i)]:
                iMin = i + 1
            elif i > iMin and A[int(i - 1)] > B[int(j)]:
                iMax = i - 1
            else:
                maxLeft = 0
                if i == 0:
                    maxLeft = B[int(j - 1)]
                elif j == 0:
                    maxLeft = A[int(i - 1)]
                else:
                    maxLeft = float(max(B[int(j - 1)], A[int(i - 1)]))
                if (m + n) % 2 == 1:
                    return maxLeft

                minRight = 0
                if i == m:
                    minRight = B[int(j)]
                elif j == n:
                    minRight = A[int(i)]
                else:
                    minRight = min(B[int(j)], A[int(i)])
                minRight = float(minRight)
                maxLeft = float(maxLeft)
                return float((maxLeft + minRight) / 2)
